### PR TITLE
Chapter 03 draft (Ruby only)

### PR DESCRIPTION
**Pull request by mattwynne**
This is my draft of the Ccumber Expressions chapter.

Seb, I'd appreciate you reading it through to see there's anything we could explain better.

I'm afraid you'll need to build it locally to read it in preview mode - we're not building branches at the moment.

```
git checkout chapter-o3-first-draft
bundle install # if you haven't done that already
bundle exec rake
open public/index.ruby.html#_expressing_yourself
```

If you're happy with it you can merge it to master (I'd use "squash and merge" as there are lots of noisy commits here)

### CODE REVIEW

**Comment by mattwynne** [comment]
Seb's feedback:

> I’ve got some concerns about section 3.5 The parameter names in the Given stepdef are the actual person’s name. Since we’re trying to emphasise that the name is parameterised, this seems confusing.
> Suggestion (06-sketch-out-parameter-type):
```
Given("{person} is located/standing {int} metre(s) from Sean") do |named_person, distance|
  named_person.move_to(distance)
end
```
Suggestion 1 (07-add-person-parameter):
```
Given("{person} is located/standing {int} metre(s) from {person}") do |named_person1, distance, named_person2|
  @lucy = named_person1
  @sean = named_person2
  @lucy.move_to(distance)
end
```
Suggestion 2 (07-add-person-parameter):
```
Given("{person} is located/standing {int} metre(s) from {person}") do |named_person1, distance, named_person2|
  @listener = named_person1
  @shouter = named_person2
  @listener.move_to(distance)
end
```
Maybe I’m over-complicating things.

File: content/03/03.02/script.ruby.adoc [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+.hear_shout.feature
+[source,gherkin]
+include::../code/ruby/02-simple-literal-expression/features/hear_shout.feature[]
+
+As Cucumber starts to execute this feature, it will come to the first step of the scenario `Given Lucy is 15 metres from Sean` and say to itself "now - do I have any step definitions that match the phrase "Lucy is 15 metres from Sean`?”"
+
+The most simple cucumber expression that would match that step is this one:
+
+----
+Lucy is located 15 metres from Sean
```
Comment (iachettifederico) [inline content/03/03.02/script.ruby.adoc]: @mattwynne this is a Step, not a Step definition

Shouldn't this be
```suggestion
Given("Lucy is located {int}m from Sean") do |distance|
```

Comment (iachettifederico) [reply content/03/03.02/script.ruby.adoc]: Also, the copy on that Step is from the old version =)

File: content/03/03.04/script.ruby.adoc [hunk 2]
```diff
@@ -0,0 +1,34 @@
+include::./title.adoc[]
+
+One common example is the problem of plurals. Suppose we want to place Lucy and Sean just 1 metre apart. This step would not match:
+
+[source, gherkin]
+----
+include::../code/ruby/04-allow-for-singular-or-plural-metres/features/hear_shout.feature[lines=3..3]
+----
+
+Because we’ve used the singular `metre` instead of the plural `metres` we don’t have a match. What a pain!
```
Comment (iachettifederico) [inline content/03/03.04/script.ruby.adoc]: I'd add a cucumber run here in order to "show, don't tell". Just a quick run, show red, come back to the code.

File: content/03/03.04/script.ruby.adoc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+include::../code/ruby/05-allow-standing-or-located-as-alternates/features/hear_shout.feature[lines=3..3]
+----
+
+...we can use this Cucumber Expression:
+[source, ruby]
+----
+include::../code/ruby/05-allow-standing-or-located-as-alternates/features/step_definitions/steps.rb[lines=3..3]
+----
+
+Now we can use either 'standing' or 'located' in our scenarios, and both will match just fine.
```
Comment (iachettifederico) [inline content/03/03.04/script.ruby.adoc]: I'd add a cucumber run here, showing that standing works

**Comment by iachettifederico** [comment]
> Seb's feedback:
> 
> > I’ve got some concerns about section 3.5 The parameter names in the Given stepdef are the actual person’s name. Since we’re trying to emphasise that the name is parameterised, this seems confusing.
> 
> Suggestion (06-sketch-out-parameter-type):
> 
> ```
> Given("{person} is located/standing {int} metre(s) from Sean") do |named_person, distance|
>   named_person.move_to(distance)
> end
> ```
> 
> Suggestion 1 (07-add-person-parameter):
> 
> ```
> Given("{person} is located/standing {int} metre(s) from {person}") do |named_person1, distance, named_person2|
>   @lucy = named_person1
>   @sean = named_person2
>   @lucy.move_to(distance)
> end
> ```
> 
> Suggestion 2 (07-add-person-parameter):
> 
> ```
> Given("{person} is located/standing {int} metre(s) from {person}") do |named_person1, distance, named_person2|
>   @listener = named_person1
>   @shouter = named_person2
>   @listener.move_to(distance)
> end
> ```
> 
> Maybe I’m over-complicating things.

I think it's a great idea to end up with 2 {person} instances.

I'd go with Lucy first, passing a `person` argument to the block, then run cucumber, show the error, then come back and implement the person matcher. Run cucumber again, now passing. 
Finally say something like "since we've converted lucy to a person pattern, we'll just convert sean as well to avoid confusion"
---
The other thing I might think about changing is the regex to accept just a word. This is not necessary, but I think it wouldn't complicate things too much (but that's just an idea)

**Comment by mattwynne** [comment]
Thanks for the feedback @iachettifederico and @sebrose. I'll get a new iteration of this done tomorrow (Thursday) and we can hopefully get it merged.

File: content/03/03.04/script.ruby.adoc (hunk 2 above)
Comment (mattwynne) [reply content/03/03.04/script.ruby.adoc]: done, thanks|!

Comment (mattwynne) [reply content/03/03.04/script.ruby.adoc]: (in 3a8d7957543991467184410b6e379e9c945f61ed)

File: content/03/03.04/script.ruby.adoc (hunk 3 above)
Comment (mattwynne) [reply content/03/03.04/script.ruby.adoc]: Thanks! Done in f733998c20e75abd309a7cab05e9ebd6a48e42da

**Comment by mattwynne** [comment]
OK, I think this is good to go.

On the parameter names, I've gone with just focussing on Lucy and capturing a single parameter. I think this is enough for illustrating the concept for now.

**Comment by mattwynne** [comment]
I'm gonna merge this, but feel free to leave more comments here in the PR if there's anything else you think we need to fix before I record the audio.

**Pull request merged**